torch/_functorch/aot_autograd.py

```diff
@@ -1,40 +1,62 @@
 import collections
 import dataclasses
-import warnings
 import itertools
+import logging
+import warnings
 from contextlib import contextmanager, nullcontext
 from dataclasses import dataclass
 from enum import Enum
-from functools import wraps, partial
+from functools import partial, wraps
 from typing import Any, Callable, Dict, List, Optional, Tuple, Union
-from torch.fx.experimental.proxy_tensor import is_sym_node
-import logging
+
+from functorch import make_fx
 
 import torch
 import torch.fx.traceback as fx_traceback
 import torch.nn as nn
 import torch.utils._pytree as pytree
 import torch.utils.dlpack
 from torch import Tensor
+from torch._dispatch.python import enable_python_dispatcher
 from torch._dynamo.utils import dynamo_timed
-from torch._subclasses import FakeTensorMode, CrossRefFakeMode, FakeTensor
+from torch._subclasses import CrossRefFakeMode, FakeTensor, FakeTensorMode
 from torch.fx import immutable_collections, Interpreter
+from torch.fx.experimental.proxy_tensor import is_sym_node, py_sym_types
 from torch.fx.experimental.symbolic_shapes import ShapeEnv
 from torch.multiprocessing.reductions import StorageWeakRef
 from torch.nn.utils import stateless
-
-from functorch import make_fx
-from torch._dispatch.python import enable_python_dispatcher
 from . import config
 from .named_members_polyfill import _named_buffers, _named_parameters
 from .partitioners import default_partition
 from torch._guards import TracingContext, DuplicateInputs
 
 log = logging.getLogger(__name__)
 
-MutationType = Enum("MutationType", ("none", "metadata_only", "data"))
+MutationType = Enum(
+    "MutationType", ("none", "metadata_only", "data", "data_and_metadata")
+)
 OutputType = Enum(
-    "OutputType", ("non_alias", "alias_of_input", "alias_of_intermediate")
+    "OutputType", (
+        # output is not an alias
+        "non_alias",
+        # output aliases an input
+        "alias_of_input",
+        # output **is** an input tensor
+        "is_input",
+        # output has a ._base tensor, which is a graph intermediate.
+        # We need to return its ._base as a graph output,
+        # so its requires_grad info is populated correctly.
+        # Instructs the runtime code to regenerate the current output
+        # from a base tensor, graph_intermediates[base_idx]
+        "alias_of_intermediate_save_as_output",
+        # Same as above; but we don't need to explicitly add its ._base
+        # as a graph output, because it already **is** a graph output.
+        "alias_of_intermediate",
+        # Same as above; but the output's ._base is **already** a user output.
```

Comment:
What's the difference between graph output and user output here, I can't really tell the difference between `alias_of_intermediate_base_is_user_output` and `alias_of_intermediate`

Reply:
Looking over the code before, it looks like `alias_of_intermediate_save_as_output` gets assigned to the first output which is an intermediate alias, and then all the other ones get `alias_of_intermediate`. And is user output is unrelated (for cases where it's already an output)